### ISSUE TITLE
Data After 40th Page Not Accessible 

### ISSUE THREAD

**Issue by LeoNerdoG** [eclipse/kapua#1155]
If user sends for more than 40 pages of data to Kapua, everything that is beyond 40th page is not accessible (look at the screenshot). 

Testflow: 
1. Login as kapua-sys
2. Connect liva Kura device to Kapua and send more than 10000 packets of data (10000 is 40 page limit in Kapua)
3. Go to Data -> Available Devices and select live device, all available metrics and click _**Query**_
4. Try to select page 41+
**At this point Kapua will try to show the page, but will get stuck in _Loading_ screen. It will not freeze, it will ust not show the content of the pages beyond 40th page.**

<img width="1438" alt="screen shot 2017-12-08 at 13 44 02" src="https://user-images.githubusercontent.com/15121999/33766608-906980e2-dc1e-11e7-9f89-e69eb3be4e0c.png">

Browsers used:
- Google Chrome (Version 62.0.3202.94 (Official Build) (64-bit))
- Safari (Version 11.0.1 (13604.3.5))
- Mozilla Firefox (57.0.1 (64-bit))
- Opera (Version 49.0.2725.39)

**Comment by lorthirk** [comment]
The error stack trace:

```
15007654 [qtp940060004-13] ERROR o.e.k.a.c.m.a.s.u.KapuaExceptionHandler - internal service error 
org.eclipse.kapua.service.datastore.internal.mediator.DatastoreException: An internal error occurred: {0}.
at org.eclipse.kapua.service.datastore.internal.MessageStoreServiceImpl.query(MessageStoreServiceImpl.java:199)
at org.eclipse.kapua.app.console.module.data.server.GwtDataServiceImpl.getMessagesList(GwtDataServiceImpl.java:335)
at org.eclipse.kapua.app.console.module.data.server.GwtDataServiceImpl.findMessages(GwtDataServiceImpl.java:322)
at org.eclipse.kapua.app.console.module.data.server.GwtDataServiceImpl.findMessagesByTopic(GwtDataServiceImpl.java:250)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at com.google.gwt.user.server.rpc.RPC.invokeAndEncodeResponse(RPC.java:569)
at com.google.gwt.user.server.rpc.RemoteServiceServlet.processCall(RemoteServiceServlet.java:208)
at com.google.gwt.user.server.rpc.RemoteServiceServlet.processPost(RemoteServiceServlet.java:248)
at com.google.gwt.user.server.rpc.AbstractRemoteServiceServlet.doPost(AbstractRemoteServiceServlet.java:62)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:841)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1650)
at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:206)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:449)
at org.eclipse.kapua.app.console.core.filter.KapuaWebFilter.executeChain(KapuaWebFilter.java:43)
at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:383)
at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1629)
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:190)
at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:219)
at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)
at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:169)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
at org.eclipse.jetty.server.Server.handle(Server.java:564)
at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:317)
at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:110)
at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
at org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:128)
at org.eclipse.jetty.util.thread.Invocable$InvocableExecutor.invoke(Invocable.java:222)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:294)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:199)
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:673)
at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:591)
at java.lang.Thread.run(Thread.java:748)
Caused by: org.eclipse.kapua.service.datastore.client.ClientException: Error: 
at org.eclipse.kapua.service.datastore.client.rest.RestDatastoreClient.restCallTimeoutHandler(RestDatastoreClient.java:777)
at org.eclipse.kapua.service.datastore.client.rest.RestDatastoreClient.query(RestDatastoreClient.java:414)
at org.eclipse.kapua.service.datastore.internal.MessageStoreFacade.query(MessageStoreFacade.java:296)
at org.eclipse.kapua.service.datastore.internal.MessageStoreServiceImpl.query(MessageStoreServiceImpl.java:197)
... 57 common frames omitted
Caused by: org.elasticsearch.client.ResponseException: GET http://es:9200/1-*/message/_search: HTTP/1.1 500 Internal Server Error
{"error":{"root_cause":[{"type":"query_phase_execution_exception","reason":"Result window is too large, from + size must be less than or equal to: [10000] but was [10250]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"1-2018-05","node":"7QDpMXJDTTWeveH89V-sZw","reason":{"type":"query_phase_execution_exception","reason":"Result window is too large, from + size must be less than or equal to: [10000] but was [10250]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."}}]},"status":500}
at org.elasticsearch.client.RestClient$1.completed(RestClient.java:354)
at org.elasticsearch.client.RestClient$1.completed(RestClient.java:343)
at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:119)
at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:177)
at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:436)
at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:326)
at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:114)
at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
... 1 common frames omitted
```

@riccardomodanese any hint?

**Comment by lorthirk** [comment]
It looks like Elasticsearch doesn't support an offset > 10000. Probably we want to investigate if different search techniques from Elasticsearch, such as [Scroll](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html) or [Search API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-search-after.html), could be implemented in our use case. Otherwise we can only limit the results window to 10000, maybe showing an alert to report that there may be more results and advice to narrow down the time window.

[Elasticsearch Docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-from-size.html)

**Comment by lorthirk** [comment]
After reading a bit more documentation, I'd say we only want to limit to 10000 the results grid, so we don't pass the 40 pages limit. I'll open another issue to investigate further the strategy to overcome the 10000 results limit.

**Comment by LeoNerdoG** [comment]
@lorthirk so we keep this issue open? Or do I close this one and you open new one?

**Comment by lorthirk** [comment]
I already opened #1378 to increase the limit, but since it has to be investigated I'd say to enforce the current limitation to 10000 in the grid now.

**Comment by lorthirk** [comment]
A fix is proposed in `fix-data_results_grid_limit` branch

**Comment by LeoNerdoG** [comment]
@lorthirk I havev verified that branch fix-data_results_grid_limit resolves this issue.

**Comment by lorthirk** [comment]
Merged in develop with e85e03c7f

**Comment by LeoNerdoG** [comment]
I have verified that this issue has been fixed.
